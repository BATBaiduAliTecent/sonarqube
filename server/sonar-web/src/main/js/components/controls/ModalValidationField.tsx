/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';

type ValidateElement = HTMLInputElement | HTMLTextAreaElement;

interface Props {
  children: (
    props: {
      className?: string;
      onBlur: () => void;
      onChange: (evt: React.SyntheticEvent<ValidateElement>) => void;
      onFocus: () => void;
      value: string;
    }
  ) => React.ReactNode;
  description?: string;
  error: string | undefined;
  label?: React.ReactNode;
  onChange: (value: string) => void;
  value: string;
}

interface State {
  editing: boolean;
  touched: boolean;
}

export default class ModalValidationField extends React.PureComponent<Props, State> {
  state: State = { editing: false, touched: false };

  handleBlur = () => this.setState({ editing: false });

  handleChange = (evt: React.SyntheticEvent<ValidateElement>) => {
    if (!this.state.touched) {
      this.setState({ touched: true });
    }
    this.props.onChange(evt.currentTarget.value);
  };

  handleFocus = () => this.setState({ editing: true });

  render() {
    const { description, error, value } = this.props;
    const { editing, touched } = this.state;
    const isValid = error === undefined;
    const showError = touched && !editing && error !== undefined;
    return (
      <div className="modal-validation-field">
        {this.props.label}
        {this.props.children({
          className: classNames({
            'has-error': showError,
            'is-valid': isValid
          }),
          onBlur: this.handleBlur,
          onChange: this.handleChange,
          onFocus: this.handleFocus,
          value
        })}
        {showError && <i className="text-danger little-spacer-left icon-alert-error" />}
        {isValid && <i className="text-success little-spacer-left icon-alert-ok" />}
        {showError && <p className="text-danger">{error}</p>}
        {description && <div className="modal-field-description">{description}</div>}
      </div>
    );
  }
}
