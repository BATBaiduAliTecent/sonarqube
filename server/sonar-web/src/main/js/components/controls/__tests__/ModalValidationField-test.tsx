/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import ModalValidationField from '../ModalValidationField';
import { change } from '../../../helpers/testUtils';

it('should display the field without any error/validation', () => {
  expect(getWrapper({ description: 'Describe Foo.' })).toMatchSnapshot();
});

it('should display the field as valid', () => {
  expect(getWrapper({ error: undefined })).toMatchSnapshot();
});

it('should display the field with an error after edition', () => {
  const onChange = jest.fn();
  const wrapper = getWrapper({ error: 'Wrong format', onChange });
  expect(wrapper.find('text-danger')).toHaveLength(0);
  change(wrapper.find('input'), 'foo');
  expect(onChange).toBeCalledWith('foo');
  expect(wrapper).toMatchSnapshot();
});

function getWrapper(props = {}) {
  return shallow(
    <ModalValidationField
      error="Is required"
      label={<label>Foo</label>}
      onChange={jest.fn()}
      value=""
      {...props}>
      {({ className, onBlur, onChange, onFocus, value }) => (
        <input
          className={className}
          onBlur={onBlur}
          onChange={onChange}
          onFocus={onFocus}
          type="text"
          value={value}
        />
      )}
    </ModalValidationField>
  );
}
