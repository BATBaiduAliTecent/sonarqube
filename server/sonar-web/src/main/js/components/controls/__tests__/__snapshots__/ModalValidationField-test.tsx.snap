// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the field as valid 1`] = `
<div
  className="modal-validation-field"
>
  <label>
    Foo
  </label>
  <input
    className="is-valid"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <i
    className="text-success little-spacer-left icon-alert-ok"
  />
</div>
`;

exports[`should display the field with an error after edition 1`] = `
<div
  className="modal-validation-field"
>
  <label>
    Foo
  </label>
  <input
    className="has-error"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <i
    className="text-danger little-spacer-left icon-alert-error"
  />
  <p
    className="text-danger"
  >
    Wrong format
  </p>
</div>
`;

exports[`should display the field without any error/validation 1`] = `
<div
  className="modal-validation-field"
>
  <label>
    Foo
  </label>
  <input
    className=""
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <div
    className="modal-field-description"
  >
    Describe Foo.
  </div>
</div>
`;
